### PR TITLE
UCP/STREAM: add UCP_REQUEST_FLAG_RECV_WAITALL support

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This is usability improvement for applications which need to wait full message to handle it. Particular receive can be used, for example, for pipelining optimization but not every application is ready to handle it properly.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6325/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6326/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6327/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6328/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6331/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6335/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6338/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6342/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@alex-mikheev @yosefe pls review

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -377,6 +377,20 @@ typedef enum {
...
 
+/**
+ * @ingroup UCP_COMM
+ * @brief Flags to define behavior of @ref ucp_stream_recv_nb function
+ *
+ * This enumeration defines behavior of @ref ucp_stream_recv_nb function.
+ */
+typedef enum {
+    UCP_RECV_FLAG_WAITALL = UCS_BIT(0)  /**< This flag requests that operation
+                                             will not be completed untill all
```
Comment (alex-mikheev) [inline src/ucp/api/ucp.h]: may be rename to UCP_STREAM_RECV_... as this flag is specific for the stream api

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -377,6 +377,20 @@ typedef enum {
...
+                                             amout of requested data is received
+                                             and placed in the user buffer. */
+} ucp_recv_flags_t;
+
```
Comment (alex-mikheev) [inline src/ucp/api/ucp.h]: ucp_stream_recv_flags_t ?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6354/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6355/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6381/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
https://github.com/openucx/ucx/issues/2328
bot:mlx:retest

**Comment by evgeny-leksikov** [comment]
@yosefe pls take a look

File: src/ucp/core/ucp_request.h [hunk 3]
```diff
@@ -38,9 +38,10 @@ enum {
     UCP_REQUEST_FLAG_SYNC                 = UCS_BIT(8),
     UCP_REQUEST_FLAG_OFFLOADED            = UCS_BIT(10),
     UCP_REQUEST_FLAG_BLOCK_OFFLOAD        = UCS_BIT(11),
-    UCP_REQUEST_FLAG_STREAM_RECV          = UCS_BIT(13),
+    UCP_REQUEST_FLAG_RECV_WAITALL         = UCS_BIT(12),
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: is it relevant only for stream? so maybe UCP_REQUEST_FLAG_STREAM_RECV_WAITALL ?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6383/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6388/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Can you add justification - why we are doing this. thanks

**Comment by evgeny-leksikov** [comment]
@shamisp pls find justification in the PR description

**Comment by shamisp** [comment]
👍

**Comment by evgeny-leksikov** [comment]
@bbenton pls review the doc update

**Comment by shamisp** [comment]
@yosefe , I didn't seen okay from @bbenton

**Pull request merged**